### PR TITLE
Introduce sanity/compatibility test for live clusters

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

**bisecting is hurting...** (hard-labored fruit this time: https://github.com/solana-labs/solana/pull/12176)

#### Summary of Changes

I think if ci time and resource is allowed, this should be run on each prs instead of nightly ci job. and it seems that running this doesn't take much time.

~~- [ ] todo what to do if the tested cluster is dead? Maybe easy turn-off knob like github's `skip-live-cluster` label?~~ (EDIT: Well, let's skip this for now? clusters are pretty stable nowadays)

~~- [ ] todo if the cluster is dead, fallback to some periodic backup of snapshot + minimum ledger?~~ (EDIT: Well, let's skip this for now? clusters are pretty stable nowadays)

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -110,7 +110,7 @@ type TransactionAccountRefCells = Vec<Rc<RefCell<Account>>>;
 type TransactionLoaderRefCells = Vec<Vec<(Pubkey, RefCell<Account>)>>;
 
 // Eager rent collection repeats in cyclic manner.
-// Each cycle is composed of <partiion_count> number of tiny pubkey subranges
+// Each cycle is composed of <partition_count> number of tiny pubkey subranges
```
Comment (ryoqun) [inline runtime/src/bank.rs:113]: let's increase my karma. :)

File: ci/buildkite-pipeline.sh [hunk 2]
```diff
@@ -125,8 +125,9 @@ wait_step() {
 }
 
 all_test_steps() {
-  command_step checks ". ci/rust-version.sh; ci/docker-run.sh \$\$rust_nightly_docker_image ci/test-checks.sh" 20
-  wait_step
+  #command_step checks ". ci/rust-version.sh; ci/docker-run.sh \$\$rust_nightly_docker_image ci/test-checks.sh" 20
```
Comment (ryoqun) [inline ci/buildkite-pipeline.sh:128]: Obviously, I need to revert these before merging!

File: ci/live-cluster-sanity.sh [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    --trusted-validator 7Np41oeYqPefeNQEHSv1UDhYrehxin3NStELsSKCT4K2 \
+    --trusted-validator GdnSyH3YtwcxFvQrVVJMm1JhTS4QVX7MFsX56uJLUfiZ \
+    --trusted-validator DE1bawNcRJB9rVm3buyMVfr8mBEoyyu73NBovf2oXJsJ \
+    --trusted-validator CakcnaRDHka2gXyfbEd2d3xsvkJkqsLw2akB3zsN1D2S \
+    --expected-genesis-hash 5eykt4UsFv8P8NJdTREpY1vzqKqZKvdpKuc147dw2N9d \
+    --expected-shred-version 64864 \
+    # for your pain-less copy-paste
+
+# UPDATE docs/src/clusters.md TOO!!
+test_with_live_cluster "testnet" \
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:140]: When backporting to v1.2, I'll remove this line.

File: docs/src/clusters.md [hunk 4]
```diff
@@ -74,20 +74,21 @@ solana config set --url https://testnet.solana.com
 
 ##### Example `solana-validator` command-line
 
+[comment]: <> (UPDATE ci/live-cluster-sanity.sh TOO!)
```
Comment (ryoqun) [inline docs/src/clusters.md:77]: TIL: https://stackoverflow.com/a/20885980

**Comment by ryoqun** [comment]
@mvines I think this pr is getting in pretty good shape. Could you review this? I changed to launch and run on adhoc GCE instance and the test duration is pretty short (~10 min) for both testnet and mainnet-beta.

File: docs/src/clusters.md [hunk 5]
```diff
@@ -36,15 +36,15 @@ solana config set --url https://devnet.solana.com
 
 ```bash
 $ solana-validator \
+    --entrypoint entrypoint.devnet.solana.com:8001 \
```
Comment (ryoqun) [inline docs/src/clusters.md:39]: Reordered for logical use order: entrypoint (contact to the cluster) => [trusted] validator (fetch genesis/snapshot) => expected-... (let's assert expected things finally)

File: docs/src/clusters.md [hunk 6]
```diff
@@ -74,20 +74,21 @@ solana config set --url https://testnet.solana.com
 
 ##### Example `solana-validator` command-line
 
+[comment]: <> (UPDATE ci/live-cluster-sanity.sh TOO!!)
```
Comment (ryoqun) [inline docs/src/clusters.md:77]: TIL: https://stackoverflow.com/a/20885980

File: ci/live-cluster-sanity.sh [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    fi
+  done
+
+  _ curl \
+    -X POST \
+    -H 'Content-Type: application/json' \
+    -d '{"jsonrpc":"2.0","id":1, "method":"validatorExit"}' \
+    http://localhost:18899
+
+  (sleep 3 && kill "$tail_pid") &
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:116]: this trick realizes `set +e`-less elegant `wait`.

File: ci/live-cluster-sanity.sh [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+  echo "--- Starting validator $cluster_label"
+
+  rm -rf cluster-sanity
+  mkdir cluster-sanity
+  ./net/ssh.sh "$instance_ip" rm -rf cluster-sanity
+  ./net/ssh.sh "$instance_ip" mkdir cluster-sanity
+
+  validator_log="$cluster_label-validator.log"
+  ./net/ssh.sh "$instance_ip" -Llocalhost:18899:localhost:18899 ./solana-validator \
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:60]: combined with `--private-rpc` and `--rpc-bind-address`, the exposure to the public internet is minimized.

File: ci/live-cluster-sanity.sh [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    if [[ (($attempts == 0)) || ! -d "/proc/$ssh_pid" ]]; then
+      on_error "start"
+    fi
+
+    sleep 3
+    echo "##### validator is starting... (until timeout: $attempts) #####"
+    show_log
+  done
+
+  echo "--- Monitoring validator $cluster_label"
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:87]: should I also add the `catchup` phase?

File: ci/live-cluster-sanity.sh [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+# UPDATE docs/src/clusters.md TOO!!
+test_with_live_cluster "testnet" \
+    --entrypoint entrypoint.testnet.solana.com:8001 \
+    --trusted-validator 5D1fNXzvv5NjV1ysLjirC4WY92RNsVH18vjmcszZd8on \
+    --trusted-validator ta1Uvfb7W5BRPrdGnhP9RmeCGKzBySGM1hTE4rBRy6T \
+    --trusted-validator Ft5fbkqNa76vnsjYNwjDZUXoTWpP7VYm3mtsaQckQADN \
+    --trusted-validator 9QxCLckBiJc783jnMvXZubK4wH86Eqqvashtrwvcsgkv \
+    --expected-genesis-hash 4uhcVJyU9pJkvQyS88uRDiswHXSCkY3zQawwpjk2NsNY \
+    # for your pain-less copy-paste
+
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:141]: I wonder it's nice to have to upload fetched snapshots to the buildkite as artifacts for reproducible testing if anything odd happens.

File: docs/src/clusters.md (hunk 6 above)
Comment (ryoqun) [reply docs/src/clusters.md:77]: I noticed docusaus can't handle this correctly, this is the reason of failing travis build.

File: ci/live-cluster-sanity.sh [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  _ curl \
+    -X POST \
+    -H 'Content-Type: application/json' \
+    -d '{"jsonrpc":"2.0","id":1, "method":"validatorExit"}' \
+    http://localhost:18899
+
+  (sleep 3 && kill "$tail_pid") &
+  kill_pid=$!
+  wait "$ssh_pid" "$tail_pid" "$kill_pid"
```
Comment (ryoqun) [inline ci/live-cluster-sanity.sh:118]: guard with `timeout N `

File: ci/live-cluster-sanity.sh [hunk 12]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
+set -e
+cd "$(dirname "$0")/.."
+
+source ci/_
+source ci/rust-version.sh stable
+
+escaped_branch=$(echo "$BUILDKITE_BRANCH" | tr -c "[:alnum:]" - | sed -r "s#(^-*|-*head-*|-*$)##g")
```
Comment (mvines) [inline ci/live-cluster-sanity.sh:8]: If `BUILDKITE_BRANCH` is empty (like if `ci/live-cluster-sanity.sh` is run locally), set `escaped_branch` to `$(whoami)` perhaps?

File: ci/live-cluster-sanity.sh [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+instance_prefix="testnet-live-sanity-$escaped_branch"
+# ensure to delete leftover cluster
+./net/gce.sh delete -p "$instance_prefix" || true
+# only bootstrap, no normal validator
+./net/gce.sh create -p "$instance_prefix" -n 0
```
Comment (mvines) [inline ci/live-cluster-sanity.sh:13]: Let's ensure the instances are shut down promptly if something goes wrong:
```suggestion
./net/gce.sh create -p "$instance_prefix" -n 0 --self-destruct-hours 1
```

File: ci/live-cluster-sanity.sh [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+      set +e
+      _ ./net/gce.sh delete -p "$instance_prefix"
+    )
+  fi
+}
+trap on_trap INT TERM EXIT
+
+_ cargo +"$rust_stable" build --bins --release
+_ ./net/scp.sh ./target/release/solana-validator "$instance_ip:."
+echo 500000 | ./net/ssh.sh "$instance_ip" sudo tee /proc/sys/vm/max_map_count > /dev/null
```
Comment (mvines) [inline ci/live-cluster-sanity.sh:28]: Instead of this, let's copy `solana-sys-tuner` in so it can set max_map_count and we verify that code path too

File: ci/live-cluster-sanity.sh [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+show_log() {
+  if find cluster-sanity/log-tail -not -empty | grep ^ > /dev/null; then
+    echo "##### new log:"
+    timeout 1 cat cluster-sanity/log-tail | tail -n 3 | cut -c 1-300 || true
+    truncate --size 0 cluster-sanity/log-tail
+    echo
+  fi
+}
+
+test_with_live_cluster() {
```
Comment (mvines) [inline ci/live-cluster-sanity.sh:48]: I find all the `./net/ssh.sh` calls in this function a little hard to follow.  I think it would be more clear if we copied a single shell script representing most/all of this function into the instance and just ran it.

File: docs/src/clusters.md [hunk 16]
```diff
@@ -36,15 +36,15 @@ solana config set --url https://devnet.solana.com
 
 ```bash
```
Comment (mvines) [inline docs/src/clusters.md:37]: The doc/ and `bank.rs` changes in here look just fine, why don't you just land those as a separate PR while we work through the `ci/` files in this PR

File: ci/live-cluster-sanity.sh [hunk 17]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
+set -e
+cd "$(dirname "$0")/.."
+
+source ci/_
+source ci/rust-version.sh stable
+
+escaped_branch=$(echo "$BUILDKITE_BRANCH" | tr -c "[:alnum:]" - | sed -r "s#(^-*|-*head-*|-*$)##g")
+instance_prefix="testnet-live-sanity-$escaped_branch"
```
Comment (t-nelson) [inline ci/live-cluster-sanity.sh:9]: I think metrics will complain about this since there won't be a database named `$instance_prefix`. (I hit similar trying to get cute with the rolling upgrades instance names)

File: ci/live-cluster-sanity.sh [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+# ensure to delete leftover cluster
+./net/gce.sh delete -p "$instance_prefix" || true
+# only bootstrap, no normal validator
+./net/gce.sh create -p "$instance_prefix" -n 0
+instance_ip=$(./net/gce.sh info | grep bootstrap-validator | awk '{print $3}')
+
+on_trap() {
+  if [[ -z $instance_deleted ]]; then
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply ci/live-cluster-sanity.sh:17]: I think it's safe to just try to delete here

File: ci/buildkite-pipeline.sh [hunk 19]
```diff
@@ -169,6 +172,12 @@ all_test_steps() {
     artifact_paths: "log-*.txt"
     agents:
       - "queue=cuda"
+  - command: "ci/live-cluster-sanity.sh"
```
[the remark this answers is not in the document]
Comment (ryoqun) [reply ci/buildkite-pipeline.sh:175]: Yeah, I think this worth on every PR. These are some reasons:

- nightly is a bit too infrequent in my opinion;
  - According to [the insights](https://github.com/solana-labs/solana/pulse), it seems that we're merging 20 PRs per business day (100 per weak/500 per month). Then, assume roughly half of it is rust (validator) related (quick guess from https://buildkite.com/solana-labs/solana/builds?branch=master&page=2). Under that numbers in mind, bisecting regressions will take about 3 steps (2 ** 3 =~ 10) in average with nightly. This is tedious in my opinion; bisecting is very effective for the very-wide window, it's not so much effective in small window. 
    - I can tolerate with hourly, but then why not every-pr? ;)
  - This doesn't make the whole CI longer from the PR author's perspective (`local-cluster` is the longest at this pipeline phase...)
- it's less ideal compared to unit-tests, but this test could serve as a smoke test around process startup, whose tests are currently particularly weak.
- Running every PR could work as a last minute sanity check in the case of hotfix.
- `live-cluster` occupies `queue=gce-deploy` which isn't so crowded compared to the `queue=default`.

[the remark this answers is not in the document]
Comment (t-nelson) [reply ci/buildkite-pipeline.sh:175]: > `live-cluster` occupies `queue=gce-deploy` which isn't so crowded compared to the `queue=default`.

I believe there are only one or two agents running `gce-deploy` ATM.  So we'll want to bump that up first.  It should just be a matter of ensuring the gcloud CLI tools are installed and pointed at the correct project, then adding a systemd service for the new agent

File: ci/live-cluster-sanity.sh (hunk 15 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh:33]: this is done!

File: ci/live-cluster-sanity.sh (hunk 14 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: done

File: ci/live-cluster-sanity.sh (hunk 13 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: thanks for tipping this nice option. I didn't know it.

File: ci/live-cluster-sanity.sh (hunk 12 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: done

File: ci/live-cluster-sanity.sh (hunk 11 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: well it turned out this is rather complicated. hint: `wait` must be shell builtin but `timeout` is just a normal command. let's skip this.

File: ci/live-cluster-sanity.sh (hunk 10 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: done by uploading snapshots only if the build failed.

File: ci/live-cluster-sanity.sh (hunk 8 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: no longer needed

File: docs/src/clusters.md (hunk 16 above)
Comment (ryoqun) [reply docs/src/clusters.md:37]: There is no strong reason to create separate PRs. I just thought it's not worth to be its own prs. `bank.rs` changes are needed to trigger `live-cluster` tests. (yeah, I could improve the `ci/buildkite-pipeline.sh`). And `docs` chagnes somewhat mentions this pr about the update notice. So separating them introduces a bit of work.

File: ci/live-cluster-sanity.sh (hunk 17 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: Fortunately, this pr doesn't use much of `net/*.sh`s. This isn't affected. Anyway, I've specifically setup a metric database for this job.

File: ci/live-cluster-sanity.sh (hunk 9 above)
Comment (ryoqun) [reply ci/live-cluster-sanity.sh]: let's skip this for now. This will increase the test time.

File: docs/src/clusters.md (hunk 6 above)
Comment (ryoqun) [reply docs/src/clusters.md]: this is now fixed.

**Comment by ryoqun** [comment]
@mvines @t-nelson I think this is good for another review in code-wise. The only remaining job is to increase build-agent for `gce-deploy`.

**Review by t-nelson — APPROVED** [review]
LGTM, now!  Rolling out more BK agents is currently blocked on https://github.com/solana-labs/solana/pull/12527, though

**Review by mvines — APPROVED** [review]
I removed the v1.2 label, I don't think we need that backport since we're about to end-of-life v1.2

**Comment by ryoqun** [comment]
just fyi: this is blocked by #12527

**Comment by ryoqun** [comment]
rebased and reverted some code commented-out for faster ci testing.

FYI: my intel agent (@carllin) tipped about another possible incompatibility of tip of master with mainnet-beta. let's test it. :)

**Comment by ryoqun** [comment]
@carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:

https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69

```
thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
```

it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.

**Comment by ryoqun** [comment]
> @carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:
> 
> https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69
> 
> ```
> thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
> ```
> 
> it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.

@jackcmay Is there something to ring in your mind. it seems that `solana-ledger-tool` still `dlopen()`s the `libsolana_bpf_loader_program.so`, which no longer build. See the `mainnet-beta/validator.log` artifact.

**Comment by jackcmay** [comment]
@ryoqun Is this a single node test?  I don't see any dependency on bpf_loader in the ledger tool.  If could be the bank doesn't have the bpf_loader builtin?

**Comment by ryoqun** [comment]
@jackcmay Yeah, this is a single node test, which connects to the actual mainnet-beta and testnet. I think there should be at least dependency on `bpf_loader` in the ledger tool. When `ledger-tool` replays the snapshot and following blocks with its `verify` subcommand and it encounters any program transactions in those blocks, it must execute those instructions. The referenced build test is failing at `ledger-tool verify`. Oddly enough, it seems `solana-validator` can process same set of transactions without problem (the test runs `solana-validator` before `solana-ledger-tool verify`).

**Comment by ryoqun** [comment]
> > @carllin it looks like I've finally managed to produce the dlopen error bug on clean ci environment, which we talked about a abit last week:
> > https://buildkite.com/solana-labs/solana/builds/32147#8893418e-7793-4a92-ad1e-7f636326eb69
> > ```
> > thread 'blockstore_processor_1' panicked at 'Failed to load: DlOpen { desc: "/home/solana/deps/libsolana_bpf_loader_program.so: cannot open shared object file: No such file or directory" }', runtime/src/native_loader.rs:117:21
> > ```
> > 
> > 
> > it looks like `solana-validator` isn't affected; only `solana-ledger-tool` is affected.
> 
> @jackcmay Is there something to ring in your mind. it seems that `solana-ledger-tool` still `dlopen()`s the `libsolana_bpf_loader_program.so`, which no longer build. See the `mainnet-beta/validator.log` artifact.

@jackcmay This will be fixed by #12759

**Comment by jackcmay** [comment]
Awesome, thanks @ryoqun @t-nelson !

**Comment by ryoqun** [comment]
> LGTM, now! Rolling out more BK agents is currently blocked on #12527, though

@t-nelson Regarding this, you and I seem to too busy to fix #12527... :) How about merging this by **making `live-cluster` run only on release branches (master, v1.4 and v1.3), not on pull request branches**? So just one or two line `skip if` in `live-cluster-sanity.sh`. So that the CI pressure should be low and we can tolerate with two `gce-deploy` agents as-is without fixing #12527 first. Or #12527 must be fixed regardless of the number of `gce-deploy`?

I think it's cost-efficient and good compromise (or sweet spot) between every-pr and nightly... ;) At least, we can avoid releasing blatant incompatible releases. (This occurred thrice recently already #12176, #12791 and #12759 and could be avoided with this pr). I want to boost our team's productivity by freeing them from manually running validators for sanity checking. And I want to free myself from digging those regressions as well. :)

**Comment by t-nelson** [comment]
Any idea what the average run time is?

**Comment by ryoqun** [comment]
> Any idea what the average run time is?

Around ~15 min real time. Because this is parallel to `local-cluster` (which takes over 20min), this shouldn't affect overall build time, assuming there is no `gce-deploy` contention. Under somewhat emergency, we can kill `master` CI and make `v1.3` and `v1.4` take the agents respectively from the 2 `gce-deploy` agents.

**Comment by t-nelson** [comment]
I fear we'll get backed up on https://buildkite.com/solana-labs/system-performance-tests/settings/schedules.  That 8hr test is going to hurt over there in Asia :wink:

**Comment by ryoqun** [comment]
> I fear we'll get backed up on https://buildkite.com/solana-labs/system-performance-tests/settings/schedules. That 8hr test is going to hurt over there in Asia wink

got it. so we're back to square one. let's the correct thing done. ;)
